CMake: Catalyst Support

etcwilde opened the pull request:
Hooking up building the zippered catalyst slices of libswiftCore.
This patch includes generating the zippered objects, as well as emitting the interfaces and binary modules when the driver supports it. Once all build environments have a driver capable of emitting the target variant module, we can remove the compiler flag check and extract the catalyst module info in the `PlatformInfo` from the same command that generates the target info for the main target module.

Fixes: rdar://144642462

etcwilde commented:
@swift-ci please test

edymtt commented on Runtimes/Core/cmake/caches/Vendors/Apple/arm64e-MacOSX.cmake:
```diff
@@ -9,8 +9,6 @@ set(CMAKE_Swift_COMPILER_TARGET "arm64e-apple-macosx${CMAKE_OSX_DEPLOYMENT_TARGE
-list(APPEND CMAKE_C_FLAGS "-darwin-target-variant" "arm64e-apple-ios13.1-macabi")
-list(APPEND CMAKE_CXX_FLAGS "-darwin-target-variant" "arm64e-apple-ios13.1-macabi")
-list(APPEND CMAKE_Swift_FLAGS "-target-variant" "arm64e-apple-ios13.1-macabi")
+set(SwiftCore_TARGET_VARIANT_COMPILER_TARGET "arm64e-apple-ios13.1-macabi" CACHE STRING "")
```
Should we allow the user to configure the deployment target (like we do for the settings above)?

etcwilde replied:
Yeah, that's reasonable.

edymtt commented on Runtimes/Core/cmake/modules/EmitSwiftInterface.cmake:
```diff
@@ -12,6 +12,16 @@ function(emit_swift_interface target)
+
+      # Emit catalyst swiftmodules and interfaces
+      if(SwiftCore_VARIANT_MODULE_TRIPLE)
+        set(variant_module_tmp_dir "${CMAKE_CURRENT_BINARY_DIR}/${target}-${SwiftCore_VARIANT_MODULE_TRIPLE}")
```
Why in the new code we are not using the module name property like for the "regular" module files, e.g.

```suggestion
        set(variant_module_tmp_dir "${CMAKE_CURRENT_BINARY_DIR}/$<TARGET_PROPERTY:${target},Swift_MODULE_NAME>-${SwiftCore_VARIANT_MODULE_TRIPLE}")
```

etcwilde replied:
`file(MAKE_DIRECTORY ...)` can't take generator expressions, so that doesn't work.
`add_custom_command(OUTPUT ...)` doesn't allow target-dependent generator expressions, so `$<TARGET_PROPERRY:...>` doesn't work.

etcwilde commented on Runtimes/Core/cmake/modules/CatalystSupport.cmake:
```diff
@@ -1,6 +1,26 @@
+  if(HAVE_Swift_VARIANT_MODULE_PATH_FLAG)
+    # Get variant module triple
+    set(module_triple_command "${CMAKE_Swift_COMPILER}" -print-target-info -target ${SwiftCore_TARGET_VARIANT_COMPILER_TARGET})
+    execute_process(COMMAND ${module_triple_command} OUTPUT_VARIABLE target_info_json)
```
no, cache entries cannot contain newlines or they get truncated.

compnerd replied:
Wow, that is ... a choice. I wonder if we can reflow the text to drop the newlines.

etcwilde replied:
Ideally we remove this in... a month or two once we get an updated driver everywhere. If you want, we can probably cache the JSON blob in a global property that we can query, but given how much CMake already queries the compilers, this is a pretty cheap check overall so I don't think it's worth doing backflips to fix when it's already supposed to go away.

etcwilde commented on Runtimes/Core/cmake/modules/CatalystSupport.cmake:
```diff
@@ -0,0 +1,6 @@
+# Add flags for generating the zippered target variant in the build
+if(SwiftCore_TARGET_VARIANT_COMPILER_TARGET)
```
Trying to stay somewhat consistent with CMake's `CMAKE_<LANG>_COMPILER_TARGET` naming. It's for the target-variant though.

compnerd replied:
Except this isn't the <LANG>_COMPILER_TARGET. I think that the slightly different name here is warranted and reasonable.

etcwilde replied:
`SwiftCore_COMPILER_VARIANT_TARGET`? The compiler is still emitting code for the target-variant into the emitted object file. It takes a full triple, but requires that the triple have a matching architecture.

etcwilde commented:
Windows failure:
```
C:\Users\swift-ci\jenkins\workspace\swift-PR-windows\sourcekit-lsp\Sources\LanguageServerProtocolJSONRPC\JSONRPCConnection.swift:370:50: error: type '(RequestID) -> Optional<any ResponseType.Type>' does not conform to the 'Sendable' protocol
368 | 
369 |     // Setup callback for response type.
370 |     decoder.userInfo[.responseTypeCallbackKey] = { (id: RequestID) -> ResponseType.Type? in
    |                                                  |- error: type '(RequestID) -> Optional<any ResponseType.Type>' does not conform to the 'Sendable' protocol

    |                                                  `- note: a function type must be marked '@Sendable' to conform to 'Sendable'

371 |       guard let outstanding = self.outstandingRequests[id] else {

372 |         logger.error("Unknown request for \(id, privacy: .public)")
```

etcwilde commented on Runtimes/Core/cmake/modules/CatalystSupport.cmake:
```diff
@@ -0,0 +1,26 @@
+if(SwiftCore_TARGET_VARIANT_COMPILER_TARGET)
+  add_compile_options(
+    "$<$<COMPILE_LANGUAGE:C,CXX>:SHELL:-darwin-target-variant ${SwiftCore_TARGET_VARIANT_COMPILER_TARGET}>"
+    "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-target-variant ${SwiftCore_TARGET_VARIANT_COMPILER_TARGET}>")
```
After further digging, `-target-variant` doesn't seem to fully work when linking with the swift driver.
The resulting dylib is not zippered, even though the objects are. Continuing to investigate.

etcwilde committed:
CMake: Include zippered catalyst slice

Add the flags to include the zippered catalyst slice if a variant target
triple is set.

etcwilde committed:
CMake: Emit catalyst modules

If the driver supports it, emit the swiftmodules and interfaces for the
target-variant slice.

etcwilde committed:
CMake: Apple: Enable catalyst slice

Add the triples to the appropriate Apple build caches, enabling zippered
catalyst builds.

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
@swift-ci please smoke test

compnerd commented on Runtimes/Core/cmake/modules/EmitSwiftInterface.cmake:
```diff
@@ -26,5 +36,18 @@ function(install_swift_interface target)
+
+    # Install catalyst interface files
+    if(SwiftCore_VARIANT_MODULE_TRIPLE)
+      install(FILES "${CMAKE_CURRENT_BINARY_DIR}/${target}-${SwiftCore_VARIANT_MODULE_TRIPLE}/${target}.swiftmodule"
```
I don't know how I feel about the `install_swift_interface` installing the swift module here. We should hoist this to the install rules for the modules, or we can have a helper to install the swiftmodules.

compnerd commented on Runtimes/Core/cmake/modules/EmitSwiftInterface.cmake:
```diff
@@ -26,5 +36,18 @@ function(install_swift_interface target)
+        DESTINATION "${SwiftCore_INSTALL_SWIFTMODULEDIR}/$<TARGET_PROPERTY:${target},Swift_MODULE_NAME>.swiftmodule")
+      install(FILES "${CMAKE_CURRENT_BINARY_DIR}/${target}-${SwiftCore_VARIANT_MODULE_TRIPLE}/${target}.swiftinterface"
+        RENAME "${SwiftCore_VARIANT_MODULE_TRIPLE}.swiftinterface"
+        DESTINATION "${SwiftCore_INSTALL_SWIFTMODULEDIR}/$<TARGET_PROPERTY:${target},Swift_MODULE_NAME>.swiftmodule")
```
copy-pasta?

compnerd replied:
Nevermind, the wrapping got me!

etcwilde replied:
No, the swift interfaces get installed into the swiftmodule directory.
The end result should be something like:

```
Swift.swiftmodule/<target>.swiftmodule
Swift.swiftmodule/<target>.swiftinterface
Swift.swiftmodule/<target>.private.swiftinterface
Swift.swiftmodule/<target-variant>.swiftmodule
Swift.swiftmodule/<target-variant>.swiftinterface
Swift.swiftmodule/<target-variant>.private.swiftinterface
```

The Destination is just the `Swift.swiftmodule/` directory. The `RENAME` sets the name to the triple.

compnerd commented on Runtimes/Core/cmake/modules/EmitSwiftInterface.cmake:
```diff
@@ -26,5 +36,18 @@ function(install_swift_interface target)
+        DESTINATION "${SwiftCore_INSTALL_SWIFTMODULEDIR}/$<TARGET_PROPERTY:${target},Swift_MODULE_NAME>.swiftmodule")
+      install(FILES "${CMAKE_CURRENT_BINARY_DIR}/${target}-${SwiftCore_VARIANT_MODULE_TRIPLE}/${target}.private.swiftinterface"
+        RENAME "${SwiftCore_VARIANT_MODULE_TRIPLE}.private.swiftinterface"
+        DESTINATION "${SwiftCore_INSTALL_SWIFTMODULEDIR}/$<TARGET_PROPERTY:${target},Swift_MODULE_NAME>.swiftmodule")
```
likewise

etcwilde committed:
CMake: Move installation of binary swiftmodule

To keep things consistent with the target and target-variant swiftmodule
files, moving the installation of the binary swift module into
`install_swift_interface`. CMake implicitly generates the target
swiftmodule, but does not know about zippered binaries, so we need to
ensure that the target-variant module is available for installation when
applicable.

etcwilde commented:
@swift-ci please smoke test

etcwilde commented:
@compnerd, moved the installation of both binary swift modules into the `install_swift_interface` function for consistency. Let me know if there's anything else you'd like.

etcwilde commented:
@swift-ci please smoke test Linux

etcwilde commented:
@swift-ci please smoke test macOS
